File: python-api/core/inference.py
import onnxruntime as ort
import numpy as np
import json
import os
from typing import List

class ONNXInferenceEngine:
    def __init__(self, model_path: str):
        # Handle different path scenarios
        possible_paths = [
            model_path,
            f"/app/model/{os.path.basename(model_path)}",
            f"model/{os.path.basename(model_path)}",
            f"../model/{os.path.basename(model_path)}"
        ]
        
        onnx_path = None
        for path in possible_paths:
            if os.path.exists(path):
                onnx_path = path
                print(f"Found ONNX model at: {path}")
                break
        
        if onnx_path is None:
            raise FileNotFoundError(f"Could not find ONNX model. Tried: {possible_paths}")
            
        self.session = ort.InferenceSession(onnx_path)
        self.input_name = self.session.get_inputs()[0].name
        self.output_name = self.session.get_outputs()[0].name
        
        # Load model info - try different paths
        info_paths = [
            '/app/model/model_info.json',
            'model/model_info.json', 
            '../model/model_info.json'
        ]
        
        for info_path in info_paths:
            if os.path.exists(info_path):
                with open(info_path, 'r') as f:
                    self.model_info = json.load(f)
                print(f"Found model info at: {info_path}")
                break
        else:
            self.model_info = {
                "input_shape": [5],
                "output_shape": [1],
                "model_type": "linear_regression",
                "framework": "sklearn"
            }
            print("Using default model info")
    
    def predict_single(self, features: List[float]) -> float:
        input_data = np.array([features], dtype=np.float32)
        result = self.session.run([self.output_name], {self.input_name: input_data})
        return float(result[0][0])
    
    def predict_batch(self, features: List[List[float]]) -> List[float]:
        input_data = np.array(features, dtype=np.float32)
        result = self.session.run([self.output_name], {self.input_name: input_data})
        return [float(pred[0]) for pred in result[0]]
    
    def get_model_info(self) -> dict:
        return self.model_info

# Try to create inference engine with better path handling
print("Creating inference engine...")
try:
    inference_engine = ONNXInferenceEngine("linear_regression.onnx")
    print("Inference engine created successfully!")
except Exception as e:
    print(f"Failed to create inference engine: {e}")
    # Create a dummy engine for testing
    class DummyEngine:
        def predict_single(self, features): return 42.0
        def predict_batch(self, features): return [42.0] * len(features)
        def get_model_info(self): return {"input_shape": [5], "output_shape": [1], "model_type": "dummy", "framework": "none"}
    
    inference_engine = DummyEngine()
    print("Using dummy inference engine")

File: rust-api/src/core/inference.rs
use anyhow::Result;
use ort::{Environment, Session, SessionBuilder, Value};
use serde_json;
use std::collections::HashMap;
use std::sync::{Arc, OnceLock};
use ndarray::{ArrayD, CowArray};
use std::fs;

static INFERENCE_ENGINE: OnceLock<InferenceEngine> = OnceLock::new();

pub struct InferenceEngine {
    session: Session,
    model_info: HashMap<String, serde_json::Value>,
}

impl InferenceEngine {
    pub fn new() -> Result<Self> {
        let environment = Arc::new(Environment::builder().build()?);
        
        // Try different model paths
        let possible_paths = [
            "/app/model/linear_regression.onnx",
            "model/linear_regression.onnx",
            "../model/linear_regression.onnx",
            "linear_regression.onnx"
        ];
        
        let mut model_path = None;
        for path in &possible_paths {
            if fs::metadata(path).is_ok() {
                model_path = Some(path);
                println!("Found ONNX model at: {}", path);
                break;
            }
        }
        
        let model_path = model_path.ok_or_else(|| {
            anyhow::anyhow!("Could not find ONNX model. Tried: {:?}", possible_paths)
        })?;
        
        let session = SessionBuilder::new(&environment)?
            .with_model_from_file(model_path)?;
        
        // Try different paths for model info
        let info_paths = [
            "/app/model/model_info.json",
            "model/model_info.json",
            "../model/model_info.json"
        ];
        
        let mut model_info = HashMap::new();
        for info_path in &info_paths {
            if let Ok(model_info_str) = fs::read_to_string(info_path) {
                if let Ok(info) = serde_json::from_str(&model_info_str) {
                    model_info = info;
                    println!("Found model info at: {}", info_path);
                    break;
                }
            }
        }
        
        // Default if not found
        if model_info.is_empty() {
            model_info.insert("input_shape".to_string(), serde_json::json!([5]));
            model_info.insert("output_shape".to_string(), serde_json::json!([1]));
            model_info.insert("model_type".to_string(), serde_json::json!("linear_regression"));
            model_info.insert("framework".to_string(), serde_json::json!("sklearn"));
            println!("Using default model info");
        }
        
        Ok(Self { session, model_info })
    }
    
    pub fn predict_single(&self, features: &[f32]) -> Result<f32> {
        let array = ArrayD::from_shape_vec(vec![1, features.len()], features.to_vec())?;
        let cow_array = CowArray::from(array);
        let input_tensor = Value::from_array(self.session.allocator(), &cow_array)?;
        let outputs = self.session.run(vec![input_tensor])?;
        
        let output_tensor = outputs[0].try_extract::<f32>()?;
        let view = output_tensor.view();
        Ok(view[[0, 0]])
    }
    
    pub fn predict_batch(&self, features: &[Vec<f32>]) -> Result<Vec<f32>> {
        let batch_size = features.len();
        let feature_size = features[0].len();
        
        let mut flat_features = Vec::with_capacity(batch_size * feature_size);
        for row in features {
            flat_features.extend_from_slice(row);
        }
        
        let array = ArrayD::from_shape_vec(vec![batch_size, feature_size], flat_features)?;
        let cow_array = CowArray::from(array);
        let input_tensor = Value::from_array(self.session.allocator(), &cow_array)?;
        let outputs = self.session.run(vec![input_tensor])?;
        
        let output_tensor = outputs[0].try_extract::<f32>()?;
        let view = output_tensor.view();
        let mut results = Vec::with_capacity(batch_size);
        for i in 0..batch_size {
            results.push(view[[i, 0]]);
        }
        Ok(results)
    }
    
    pub fn get_model_info(&self) -> &HashMap<String, serde_json::Value> {
        &self.model_info
    }
}

pub fn get_inference_engine() -> Result<&'static InferenceEngine> {
    INFERENCE_ENGINE.get_or_init(|| {
        match InferenceEngine::new() {
            Ok(engine) => {
                println!("Inference engine created successfully!");
                engine
            }
            Err(e) => {
                println!("Failed to create inference engine: {}", e);
                panic!("Could not initialize inference engine: {}", e);
            }
        }
    });
    Ok(INFERENCE_ENGINE.get().unwrap())
}